src/common/chunked.ts

```diff
@@ -296,28 +310,90 @@ export class ChunkedUploader {
       throw firstError;
     }
 
+    const finalizeResponse = await this.finalizeUpload(
+      uploadId,
+      dataItemByteCount,
+      dataItemOpts?.paidBy,
+      combinedSignal,
+    );
+
+    emitter.emit('upload-success');
+    return finalizeResponse;
+  }
+
+  private toGiB(bytes: ByteCount): ByteCount {
+    return bytes / 1024 ** 3;
+  }
+
+  private async finalizeUpload(
+    uploadId: string,
+    dataItemByteCount: ByteCount,
+    paidBy?: string | string[],
+    signal?: AbortSignal,
+  ): Promise<TurboUploadDataItemResponse> {
+    // Wait up to 1 minute per GiB of data for the upload to finalize
+    const fileSizeInGiB = Math.ceil(this.toGiB(dataItemByteCount));
+    const defaultMaxWaitTime = fileSizeInGiB;
+    const maxWaitTimeMs =
+      this.maxFinalizationWaitTimeMs ?? defaultMaxWaitTime * 60 * 1000;
+
+    const waitPerAttempt =
```

Comment:
Is this dividing the clock time allotted per each async finalize among its max attempts? I don't really love the approach if so. Instead I think you'd want to check against elapsed time. If the first attempt takes 2 seconds out a max 5 seconds of wait time, that would be fine, but as coded would fail for taking more than 1 second. Maybe I'm misinterpreting how this is going to be used...

Reply:
changed up here: 992a566